metricplugin/tcp_client.go

```diff
@@ -0,0 +1,403 @@
+package metricplugin
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-msgio"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// A client, from the perspective of the server.
+type tcpClient struct {
+	// The address of the client.
+	remote net.Addr
+
+	// The TCP connection.
+	conn net.Conn
+
+	// A 4-byte, big-endian frame-delimited writer.
+	writer msgio.WriteCloser
+
+	// A 4-byte, big-endian frame-delimited reader.
+	reader msgio.ReadCloser
+
+	// Buffered channel for outgoing messages.
+	// This dictates backpressure.
+	msgOut chan outgoingTCPMessage
+
+	// A lock used to coordinate shutting down.
+	closingLock sync.Mutex
+	// Closed after a shutdown has been initiated.
+	closed chan struct{}
+}
+
+func newClient(c net.Conn, closing <-chan struct{}) (*tcpClient, error) {
+	client := &tcpClient{
+		remote: c.RemoteAddr(),
+		conn:   c,
+		writer: msgio.NewWriter(c),
+		reader: msgio.NewReader(c),
+		closed: make(chan struct{}),
+		msgOut: make(chan outgoingTCPMessage, 100),
```

Comment:
Is the buffer of 100 arbitrary or is there a reasoning to this?

Reply:
Mostly arbitrary! The whole backpressure situation is annoying to deal with, and I haven't exactly figured out why it's happening at all, even sometimes in low-traffic situations.

About the buffer size: An unbuffered channel crashes the plugin in case of a slow subscriber (still not sure if the plugin is too slow sending out responses or the clients are too slow receiving them. The Rust client should be extremely lightweight, but also uses some async stuff which might be difficult to reason about...). A too small buffer potentially drops too many messages in bursty situations. I don't have metrics on the average queue size, which would be nice...

This works mostly okay for our setup, maybe others want to adjust this, we'll have to see. I didn't want to expose too many knobs to tweak (without any idea how to tweak them).